### PR TITLE
refactor(libqtile): rename border(width) to border_{color,width}

### LINKED ISSUE #2825: Inconsistent use of border/border_color and borderwidth/border_width
# Issue description

Title - some of the widgets are not consistent with other configurables with these config options. See [this comment](https://github.com/qtile/qtile/pull/2675#issuecomment-927275743) for more details.

It would be nice to make these consistent with other configurables, perhaps adding a migration and deprecation notice for a version or two.

# Qtile version

0.18.1

### PR DESCRIPTION
**Pull request by ayntgl**
- The `migrate` script is not tested; ideally, the `test/scripts` directory should have tests for migrate script.

Closes #2825

### CODE REVIEW

**Comment by m-col** [comment]
This is fantastic, migration and all, thank you for this!

My first thought is does this cover every single instance of `borderwidth` and `bordercolor`? A grep would be helpful for catching all of them, e.g. `backend/wayland/window.py` also uses these.

I'm also wondering how disruptive this might be. While some people hopefuly use `qtile migrate`, some don't and may get problems. I wonder how doable a phased deprecation might be, e.g. by using a keyword argument to any `Configurable` class options that catch the old versions and log a deprecation notice, and maybe keeping old versions in any `cmd_` signatures to do the same. Maybe that's overkill/unnecessary? What do you think @elParaguayo?

Lastly, please could you add a note to the changelog under the config breakage section

**Comment by elParaguayo** [comment]
Nice!

My personal feeling on `migrate` scripts is that I don't particularly like them. I'm all in favour of trying to prevent breakages for users but:
- I'm sceptical as to how many users actually use the script
- I have reservations about how effective our scripts are (i.e. do we actually cover all ways a change may be required?)
- Depending on the change, the tests can be complicated to write which is frustrating given the above two points
- It looks like `bowler` is not supported past python 3.8 so may be incompatible with some users' configs.

For those reasons, I prefer the phased deprecation as per @m-col 's post above. We've got examples of this already elsewhere in the codebase.

**Comment by m-col** [comment]
> EDIT: I also appreciate that my comment in the issue that this closes (https://github.com/qtile/qtile/issues/2825) said we needed a migrate script. My views have evolved since then...!

Ha, nice.

> It looks like bowler is not supported past python 3.8 so may be incompatible with some users' configs.

To be precise, bowler is no longer adding language features introduced after 3.8's walrus operator. It does work on more recent python versions and will probably continue to work for quite some time, but won't consider e.g. the newer `match` syntax. But yes, fair point. Agreed - phased deprecation would be good.

**Comment by elParaguayo** [comment]
> Thank you for the feedback, @m-col @elParaguayo! Do you mind if you can point me where exactly example of phased deprecation is in the codebase?

Here's an example where we moved from a positional argument to a new parameter in `defaults`:
https://github.com/qtile/qtile/blob/b7b458073941264b2f5a84c0b49f4ed316679a9b/libqtile/widget/launchbar.py#L88-L103

**Comment by elParaguayo** [comment]
> > It looks like bowler is not supported past python 3.8 so may be incompatible with some users' configs.
> 
> To be precise, bowler is no longer adding language features introduced after 3.8's walrus operator. It does work on more recent python versions and will probably continue to work for quite some time, but won't consider e.g. the newer `match` syntax. But yes, fair point. Agreed - phased deprecation would be good.

Yes - that's fair. It's not that the code doesn't work on later versions but that they won't/can't update the grammar parsing for new features.

File: libqtile/popup.py [hunk 1]
```diff
@@ -67,6 +68,13 @@ def __init__(
         self.add_defaults(Popup.defaults)
         self.qtile = qtile
 
+        if "border" in config:
+            logger.warning(
```
Comment (elParaguayo) [inline libqtile/popup.py:72]: Make sure `logger` is imported:
``` python
from libqtile.log_utils import logger
```

File: libqtile/widget/groupbox.py [hunk 2]
```diff
@@ -50,17 +50,26 @@ def __init__(self, **config):
         self.add_defaults(base.PaddingMixin.defaults)
         self.add_defaults(base.MarginMixin.defaults)
 
+        if "borderwidth" in config:
+            logger.warning(
```
Comment (elParaguayo) [inline libqtile/widget/groupbox.py:54]: Make sure `logger` is imported:
``` python
from libqtile.log_utils import logger
```
You may need to add this to the other files you changed.

File: libqtile/widget/groupbox.py [hunk 3]
```diff
@@ -36,11 +36,12 @@
 
 from libqtile import bar, hook
 from libqtile.widget import base
+from libqtile.log_utils import logger
```
Comment (elParaguayo) [inline libqtile/widget/groupbox.py:39]:
```suggestion
from libqtile.log_utils import logger
from libqtile.widget import base
```

**Comment by elParaguayo** [comment]
OK - I've just tried this patch and there are a few bugs. I'll post some comments later.

File: libqtile/popup.py [hunk 4]
```diff
@@ -67,6 +68,13 @@ def __init__(
...
+                'The "border" argument is obsolete and deprecated.',
+                'As a substitute, you may use "border_color" argument instead.',
+            )
```
Comment (elParaguayo) [inline libqtile/popup.py:75]:
```suggestion
            logger.warning(
                'The "border" argument is obsolete and deprecated. '
                'As a substitute, you should use "border_color" argument instead.'
            )
```
The old `,` at the end of the first line meant you were passing two arguments to `logger.warning` when actually it should be a single string.

Also "may" -> "should".

This should be updated for the other messages.

Comment (elParaguayo) [reply libqtile/popup.py:75]: We should do tests too to check that the log messages are triggered when using deprecated configs.

An example is here:
https://github.com/qtile/qtile/blob/d827edd7b04c2fa95827c33aad3146794dfc39b2/test/widgets/test_launchbar.py#L31-L38

I don't think you'll need the monkeypatching, just initialise the widget with the deprecated parameters and check `caplog.records` for the messages.

I can help if you get stuck with these.

File: libqtile/scripts/migrate.py [hunk 5]
```diff
@@ -130,6 +138,8 @@ def windowtogroup_groupName_argument(funcname, query):  # noqa: N802
     tile_master_windows_rename,
     threaded_poll_text_rename,
     pacman_to_checkupdates,
+    border_to_border_color,
+    borderwidth_to_border_width,
```
Comment (elParaguayo) [inline libqtile/scripts/migrate.py:142]: If we're adding migrate scripts, we should test them.

Have a look at https://github.com/qtile/qtile/blob/master/test/test_migrate.py for examples.

File: CHANGELOG [hunk 6]
```diff
@@ -2,6 +2,7 @@ Qtile x.xx.x, released XXXX-XX-XX:
     !!! Config breakage !!!
         - lazy.qtile.display_kb() no longer receives any arguments. If you passed it any arguments
           (which were ignored previously), remove them.
+        - To maintain consistency with other widgets, all of the instances of `border` parameter are renamed to `border_color` and `borderwidth` parameter to `border_width`. To easily to replace the deprecated parameter names, `qtile migrate` may be ran to replace instances of both of these parameters.
```
Comment (elParaguayo) [inline CHANGELOG:5]:
```suggestion
        - To maintain consistency with other widgets, all of the instances of `border`
          parameter are renamed to `border_color` and `borderwidth` parameter to
          `border_width`. To easily to replace the deprecated parameter names, 
          `qtile migrate` may be ran to replace instances of both of these parameters.
```

Splitting up long line.

**Review by elParaguayo — CHANGES_REQUESTED** [review]

**Pull request closed without merging**